### DISCUSSION TITLE
Wondering how others are using keepassxc-cli.

### DISCUSSION THREAD

**Discussion by Entreprenerdz** [keepassxreboot/keepassxc#7936] [Q&A]
I've taken a look at the CLI and my high hopes of scripting EVERYTHING were quickly dashed.... I'm not sure if the CLI is fairly new and slowly having new features added or if I'm just not trying to use it as it was intended so I thought I'd reach out to see what others are using it for?

Furthermore, how are others handling the password as I doubt one would enter it every time.

I have a Yubikey which is great but a single point of major FAILURE should I ever lose it, break it, etc.  I also use KeepassXD, which does not have Yubikey integration so I've actually just gotten by without using my phone for online activities involving passwords.

I thought it would be great if I could schedule a script to run weekly to 1. Unlock the kdbx. 2. Remove the Yubikey protection.  3. Change the password to a very strong pwd.  4.  Save a backup to protect against Yubikey loss/failure.  5. Export a smaller subset of critical pwds, with a less strong pwd for my phone to access.  6. Reinstate Yubikey protection, change the pwd back to original, save.

Looking forward to hearing what others have built/done with the CLI.  Cheers.

**Comment by droidmonkey** [comment]
You should use keeshare instead. Place all credentials you want to have shared with mobile in the shared group with a strong password protecting it. Set it to export only or sync if you will make changes on your mobile device. Save the share to icloud. 

https://keepassxc.org/docs/KeePassXC_UserGuide.html#_database_sharing_with_keeshare

**Comment by XHawk87** [comment]
I sync my password database between all of my devices using Syncthing. In order to avoid conflicts, I keep a separate primary copy for each device that it alone is allowed to write to, and then I use the CLI to automatically merge the other copies into the primary when there are new changes detected.

I thought about using KeeShare, but it wouldn't solve my main issue of conflicted copies when multiple devices try to write to the same file. It would just be the .share file that is conflicted rather than the .kdbx. Also, I wanted to sync all of my passwords not just some of them, so it seemed like a bit of a roundabout way of doing that. Though it seems like a good solution if I wanted to share a subset of passwords with someone else at some point. 

One issue I did come across is that `keepasxc-cli open` seems to be intended to be used interactively rather than automated with a script. It seems that the password prompt and all subsequent prompts will consume all input from the pipe buffer, not just the next line, and discard the rest. I had to create an expect wrapper to wait until the CLI is ready for each line before supplying the next one so that it behaves more as you might expect it to:

Disclaimer: If anyone decides to use this, use at your own risk. I am not an expert on TCL/Expect, and this might have errors and I am sure a better approach could have been taken by someone more experienced.
```
#!/usr/bin/expect -f
#exp_internal 1

# initialize global variables
set timeout -1
set database_name ""

# Perform cleanup tasks
proc wrapper_exit {reason} {
    send "exit\n"
    send_user "exit\n$reason\n"
    exit
}

proc database_exit {reason} {
    send_user "\n$reason\n"
    exit
}

# Wait until the CLI is ready for next input
proc wait_for_prompt {} {
    global database_name
    expect {
        eof {
            [database_exit "Database closed unexpectedly while waiting for prompt"]
        }
        -exact "\n${database_name}> " { return 0 }
    }
}

# Supply password when the CLI asks for it
proc authenticate {} {
    global database_name
    expect -exact "Enter password to unlock "
    expect {
        eof {
            [database_exit "Database closed unexpectedly while waiting for password prompt"]
        }
        -regexp "(.+): " {
            set database_name [file tail $expect_out(1,string)]
            unset expect_out
            expect_user -regexp "(.*?)\n"
            send "$expect_out(1,string)\n"
            unset expect_out
            expect {
                eof {
                    [database_exit "Database closed unexpectedly after sending password"]
                }
                "Error" {
                    [database_exit "Database closed unexpectedly due to error after sending password"]
                }
                -exact "${database_name}> " { return 0 }
            }
        }
    }
}

# main loop for getting lines from user and supplying to spawned process
proc main_loop {} {
    expect_user {
        eof {
            [wrapper_exit "Closing the database as there is no more user input"]
        }
        -regexp "(.*?)\n" {
            if {$expect_out(1,string) eq "exit"} { [wrapper_exit "Completed!"] }
            send "$expect_out(1,string)\n"
            unset expect_out
            wait_for_prompt
            exp_continue
        }
    }
}

# spawn process and start main loop
spawn keepassxc-cli open {*}$argv
authenticate
main_loop
```
This is quite a brittle solution, any changes in the referenced CLI text will break it. It might be better to address this in the CLI code instead.
